### PR TITLE
Put symhacks in crypto.h; dont edit generated file.

### PR DESCRIPTION
**Pull request by richsalz**

### CODE REVIEW

File: include/openssl/crypto.h [hunk 1]
```diff
@@ -24,6 +24,7 @@
 # include <openssl/opensslv.h>
 # include <openssl/ossl_typ.h>
 # include <openssl/opensslconf.h>
+# include <openssl/symhacks.h>
```
Comment (levitte) [inline include/openssl/crypto.h]: If someone include `openssl/cryptoerr.h` only, this won't help.  You should look in `crypto/cpt_err.h`...  or, for that matter, `crypto/params.c` in certain PRs :wink:

Comment (richsalz) [reply include/openssl/crypto.h]: I can think of two alternatives

- Change mkerr to put it in EVERY generated err file
- Change mkerr to have a special flag to HACK it into just that one file.

The latter is too gross to consider.

If you have a better idea that'd be great.

Editing auto-generated files is horrific.  That's the whole reason why I rewrote mkerr originally :)

Comment (levitte) [reply include/openssl/crypto.h]: > * Change mkerr to put it in EVERY generated err file

That seems like the better alternative.

Comment (richsalz) [reply include/openssl/crypto.h]: okay, done.

File: util/mkerr.pl [hunk 2]
```diff
@@ -455,6 +453,10 @@ sub help
 #ifndef HEADER_${lib}ERR_H
 # define HEADER_${lib}ERR_H
 
+# ifndef HEADER_SYMHACKS_H
```
Comment (levitte) [inline util/mkerr.pl]: Side note: Not really sure why that's needed...

(no, you don't have to change anything...)

**Comment by levitte** [comment]
Side note: At some point, we should really change all the header file guards.  `HEADER_whatever_H` is waaaaay too common.  That's for another PR, though

**Comment by levitte** [comment]
Merged.

2a6a56073ce2ccac3e8c83b5434c6472ab6354e4 Add regenerated header files
b53c4fe3f92e3d2c5bd9fca1a171cd24f66ef14d util/mkerr.pl: Add an inclusion of symhacks.h in all error files

**Comment by mspncp** [comment]
> Side note: At some point, we should really change all the header file guards. `HEADER_whatever_H` is waaaaay too common. That's for another PR, though

For the records: the wish has come true, see #9274

@levitte do I get your permission to cherry-pick this pull request to 1.1.1? (I tried it, it applies without conflicts.) The fact that it's missing in 1.1.1 causes conflicts when cherry-picking 755b50599854734f4c70c11fe7899d875acf6db2.

**Comment by levitte** [comment]
Yes, do cherry pick to 1.1.1. @romen, do you agree?

**Comment by mspncp** [comment]
> Yes, do cherry pick to 1.1.1. @romen, do you agree?

Thanks. I noticed that I will have to raise a separate pull request for 1.1.1 anyway, because only b53c4fe3f92e3d2c5bd9fca1a171cd24f66ef14d cherry-picks cleanly, but the generated files 2a6a56073ce2ccac3e8c83b5434c6472ab6354e4 don't.

BTW: what is the correct way to rebuild the files? I guess I need to run

```
perl -I. util/mkerr.pl -internal -rebuild
```

and then discard all files for which only the copyright year changed? Or is there a better way to do it?

**Comment by levitte** [comment]
Why do you need to rebuild anything? Rebuilding means renumbering, doesn't it?

**Comment by levitte** [comment]
... Or I dunno. I won't be able to check the technical details before tomorrow. Traveling today.

**Comment by mspncp** [comment]
> ... Or I dunno. I won't be able to check the technical details before tomorrow. Traveling today.

No problem, I'm not in a hurry. Maybe @richsalz can give me advice whether I should regenerate the files and how to do it. Anyway, I'll keep only the symhacks changes and discard any renumbering.

```C
# ifndef HEADER_SYMHACKS_H
#  include <openssl/symhacks.h>
# endif
```

**Comment by richsalz** [comment]
Rebuilding the files should NOT cause renumbering; that's why state is kept in openssl.txt

**Comment by romen** [comment]
> Yes, do cherry pick to 1.1.1. @romen, do you agree?

Probably worthless at this point, but yes!

**Comment by mspncp** [comment]
> Rebuilding the files should NOT cause renumbering; that's why state is kept in openssl.txt

What's the correct command line to do the renumbering? (i.e how did _you_ do it?)

**Comment by mspncp** [comment]
Sorry, I meant 'rebuilding'.

**Comment by richsalz** [comment]
```
make ERROR_REBUILD=-rebuild  errors
```

**Comment by mspncp** [comment]
Pull request for backport created, see #9281.

**Pull request closed without merging**